torchao/quantization/GPTQ.py

```diff
@@ -0,0 +1,579 @@
+                    *[
+                        (
+                            x.values
+                            if isinstance(x, MultiInput)
+                            else [x] * multi_input_count
+                        )
+                        for x in flat_args
+                    ]
+                )
+            )
+        else:
+            transposed_args = [flat_args]
+        outputs = []
+
+        # check whether we apply GPTQ to this module
+        quantize_linear = (
+            (target == aten.linear.default)  # if its a linear
+            and id(args[1]) in self.id_to_name  # and if we know the layer name
+            and not skip_quant  # and if we weren't told to skip quantization
+            # and if the skip_layer_func doesn't say we should skip
+            and not (self.skip_layer_func is not None and self.skip_layer_func(args[1]))
+        )  # then we will quantize this linear layer/weight
+
+        if quantize_linear:  # instantiate variables for GPTQ
+            H = 0
+            total_batches = 0
+
+        for inp in transposed_args:
+            inp = tensors_to_cuda(inp)
+            cur_args, cur_kwargs = tree_unflatten(inp, spec)
+
+            if (
+                quantize_linear
+            ):  # calculate H instead of output (will run the linear eventually with updated weight)
+                x = cur_args[0].float()
+                shape = x.shape
+                n = 1 if len(shape) == 2 else shape[0]
+                H *= total_batches / (total_batches + n)
+                total_batches += n
+                x = ((2 / total_batches) ** (1 / 2)) * x.reshape(
+                    -1, shape[-1]
+                ).t().float()
+                H += x.matmul(x.t())
+            else:
+                # get output if its not a linear
+                out = super().call_function(target, cur_args, cur_kwargs)
+
+                if isinstance(out, torch.Tensor):
+                    outputs.append(out.cpu())
+                else:
+                    outputs.append(out)
+
+        if quantize_linear:
+            mod_fqn = ".".join(self.id_to_name[id(args[1])].split(".")[:-1])
+            W = args[1].to(H.device)
+            Q, DQ, qparams = self.faster_quant(H, W.detach())
+            print(mod_fqn)
+            names_and_values_dict = self.make_names_and_values_dict_func(Q, qparams)
+
+            # delete old weight
+            if mod_fqn + ".weight" in self.new_state_dict:
+                self.new_state_dict.pop(mod_fqn + ".weight")
+            if len(args) > 2:
+                self.new_state_dict[mod_fqn + ".bias"] = args[2]
+            for name, value in names_and_values_dict.items():
+                self.new_state_dict[mod_fqn + "." + name] = value
+
+            # run linear with new weight to get corrected output
+            new_out = self.call_function(
+                target, (args[0], DQ, *args[2:]), kwargs, skip_quant=True
+            )
+
+            if self.debug:
```

Comment:
This seems generic enough that maybe we'd want it be decoupled?

Reply:
sure, this is copy pasting btw, we can follow up.

by decouple do you mean abstract this part so that others can use it as well right